cbmc/proofs/IotMqtt_SubscribeAsync/IotMqtt_SubscribeAsync_harness.c

```diff
@@ -0,0 +1,107 @@
+{
+  IotTaskPoolError_t error;
+
+  /* _IotMqtt_ScheduleOperation asserts this */
+  __CPROVER_assume(error != IOT_TASKPOOL_BAD_PARAMETER &&
+                   error != IOT_TASKPOOL_ILLEGAL_OPERATION);
+  return error;
+}
+
+/**
+ * _IotMqtt_NextPacketIdentifier calls Atomic_Add_u32, which receives
+ * a volatile variable as input. Thus, CBMC will always consider that
+ * Atomic_Add_u32 will operate over nondetermistic values and raises
+ * a unsigned integer overflow failure. However, developers have reported
+ * that the use of this overflow is part of the function implementation.
+ * In order to mirror _IotMqtt_NextPacketIdentifier behaviour and avoid
+ * spurious alarms, we stub out this function to always
+ * return a nondetermistic odd value.
+ */
+uint16_t _IotMqtt_NextPacketIdentifier( void ) {
+  uint16_t id;
+
+  /* Packet identifiers will follow the sequence 1,3,5...65535,1,3,5... */
+  __CPROVER_assume(id & 0x0001 == 1);
+
+  return id;
+}
+
+void IotListDouble_RemoveAllMatches( const IotListDouble_t * const pList,
+				     bool ( *isMatch )( const IotLink_t * const pOperationLink, void * pCompare ),
+				     void * pMatch,
+				     void ( *freeElement )( void * pData ),
+				     size_t linkOffset )
+{
+  free_IotMqttSubscriptionList( pList );
+  allocate_IotMqttSubscriptionList( pList, SUBSCRIPTION_COUNT_MAX-1 );
+}
+
+void _IotMqtt_RemoveSubscriptionByTopicFilter( _mqttConnection_t * pMqttConnection,
+                                               const IotMqttSubscription_t * pSubscriptionList,
+                                               size_t subscriptionCount )
+{
+  free_IotMqttSubscriptionList( &( pMqttConnection->subscriptionList ) );
+  allocate_IotMqttSubscriptionList( &( pMqttConnection->subscriptionList ), subscriptionCount );
+}
+
+void harness()
+{
+  /* assume a valid MQTT connection */
+  IotMqttConnection_t mqttConnection = allocate_IotMqttConnection(NULL);
+  __CPROVER_assume(mqttConnection != NULL);
+  __CPROVER_assume(mqttConnection->pNetworkInterface != NULL);
+  __CPROVER_assume( IS_STUBBED_NETWORKIF_SEND( mqttConnection->pNetworkInterface ) );
+  __CPROVER_assume( IS_STUBBED_NETWORKIF_DESTROY( mqttConnection->pNetworkInterface ) );
+  ensure_IotMqttConnection_has_lists(mqttConnection);
+  __CPROVER_assume(valid_IotMqttConnection(mqttConnection));
+
+  /* assume unconstrained inputs */
+  size_t subscriptionCount;
+  __CPROVER_assume(subscriptionCount < SUBSCRIPTION_COUNT_MAX);
+  IotMqttSubscription_t *pSubscriptionList = allocate_IotMqttSubscriptionArray(NULL, subscriptionCount);
+  __CPROVER_assume(valid_IotMqttSubscriptionArray(pSubscriptionList, subscriptionCount));
+  uint32_t flags;
+  IotMqttCallbackInfo_t *pCallbackInfo = malloc_can_fail(sizeof (*pCallbackInfo) );
+
+  /* output */
+  IotMqttOperation_t pSubscribeOperation = allocate_IotMqttOperation(NULL, mqttConnection);
+  if(pSubscribeOperation != NULL)
+  {
+    __CPROVER_assume(valid_IotMqttOperation(pSubscribeOperation));
```

Comment:
This is an output, how much of this struct do we have to allocate ourselves?  Is it really necessary to assume the operation is a valid operation before the SubAsync function overwrites it?

Reply:
Good catch! We actually don't even need to assume it's valid here, I removed it.